[Workers] add upstash integration (#10168)
cloudflare/cloudflare-docs: content/workers/databases/native-integrations/upstash.md

---
pcx_content_type: concept
title: Upstash
---
## Connecting to Upstash

[Upstash](https://upstash.com/) is a serverless database with Redis and Kafka API. Upstash also offers QStash, a task queue/scheduler designed for the serverless. 

### Set up an integration with Upstash Redis:

1. You need an existing Upstash database to connect to. [Create a Upstash database](https://docs.upstash.com/redis#create-a-database) or [load data from an existing database to Upstash](https://docs.upstash.com/redis/howto/connectclient).
2. Insert some data to your Upstash database. You can add data to your Upstash database in two ways:
    - Use the CLI directly from your Upstash console.
    - Alternatively, install [redis-cli](https://redis.io/docs/getting-started/installation/) locally and run the following commands. 
    ```sh
    ➜ set GB "Ey up?"
    OK
    ➜ set US "Yo, what’s up?"
    OK
    ➜ set NL "Hoi, hoe gaat het?"
    OK
    ```

> 3. Add the Upstash Redis integration to your Worker:
>     - Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
>     - In **Account Home**, select **Workers & Pages**.
>     - In **Overview**, select your Worker.
>     - Select **Settings** > **Integrations** > **Upstash Redis**. 
>     - Follow the setup flow, selecting the database created in step 1.
> 4. In your Worker, install the `@upstash/redis`, a HTTP client to connect to your database and start manipulating data:
>     ```sh
>     $ npm install @upstash/redis
>     ```

Reviewer: ```suggestion
3. Add the Upstash Redis integration to your Worker:

    a. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
    b. In **Account Home**, select **Workers & Pages**.
    c. In **Overview**, select your Worker.
    d. Select **Settings** > **Integrations** > **Upstash Redis**. 
    e. Follow the setup flow, selecting the database created in step 1.
    
4. In your Worker, install the `@upstash/redis`, a HTTP client to connect to your database and start manipulating data:

    ```sh
    $ npm install @upstash/redis
    ```

```

> 5. The following example shows how to make a query to your Upstash database in a Worker. The credentials needed to connect to Upstash have been automatically added as secrets to your Worker through the integration.
>   ``` js
>   import { Redis } from "@upstash/redis/cloudflare";
> 
>   export default {
>     async fetch(request, env) {
>       const redis = Redis.fromEnv(env);
> 

Reviewer: ```suggestion
5. The following example shows how to make a query to your Upstash database in a Worker. The credentials needed to connect to Upstash have been automatically added as secrets to your Worker through the integration.

  ``` js
  ---
  filename: Worker code
  ---
  import { Redis } from "@upstash/redis/cloudflare";

  export default {
    async fetch(request, env) {
      const redis = Redis.fromEnv(env);

```

      const country = request.headers.get("cf-ipcountry");
      if (country) {
        const greeting = await redis.get(country);
        if (greeting) {
          return new Response(greeting);
        }
      }

      return new Response("Hello What's up!");
    },
  };
  ```

To learn more about Upstash, refer to the [Upstash documentation](https://docs.upstash.com/redis).

### Set up an integration with Upstash Kafka:
1. Create a [Kafka cluster and topic](https://docs.upstash.com/kafka)
2. Add the Upstash Kafka integration to your Worker:

>     - Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
>     - In **Account Home**, select **Workers & Pages**.
>     - In **Overview**, select your Worker.
>     - Select **Settings** > **Integrations** > **Upstash Kafka**. 
>     - Follow the setup flow, selecting the cluster and topic.
> 3. In your Worker, install `@upstash/kafka`, a HTTP/REST based Kafka client: 
>     ```sh
>     $ npm install @upstash/kafka
>     ```

Reviewer: ```suggestion
    a. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
    b. In **Account Home**, select **Workers & Pages**.
    c. In **Overview**, select your Worker.
    d. Select **Settings** > **Integrations** > **Upstash Kafka**. 
    e. Follow the setup flow, selecting the cluster and topic.
    
3. In your Worker, install `@upstash/kafka`, a HTTP/REST based Kafka client:

    ```sh
    $ npm install @upstash/kafka
    ```

```

4. Use the [upstash-kafka](https://github.com/upstash/upstash-kafka/blob/main/README.md) JavaScript SDK to send data to Kafka.
Check out an [example](https://docs.upstash.com/kafka/real-time-analytics/realtime_analytics_serverless_kafka_setup#option-1-cloudflare-workers) published by Upstash. Replace `url`, `username` and `password` with the variables set by the integration. 

> ### Set up an integration with Upstash QStash:
> 1. Configure the [publicly available HTTP endpoint](https://docs.upstash.com/qstash#1-public-api) that you want to send your messages to. 

Reviewer: ```suggestion
## Upstash QStash

To set up an integration with Upstash QStash:

1. Configure the [publicly available HTTP endpoint](https://docs.upstash.com/qstash#1-public-api) that you want to send your messages to.

```

> 2. Add the Upstash QStash integration to your Worker:
>     - Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
>     - In **Account Home**, select **Workers & Pages**.
>     - In **Overview**, select your Worker.
>     - Select **Settings** > **Integrations** > **Upstash QStash**. 
>     - Follow the setup flow..

Reviewer: ```suggestion
2. Add the Upstash QStash integration to your Worker:

    a. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
    b. In **Account Home**, select **Workers & Pages**.
    c. In **Overview**, select your Worker.
    d. Select **Settings** > **Integrations** > **Upstash QStash**. 
    e. Follow the setup flow.
    
```
Was there supposed to be something after follow the setup flow?

> 3. In your Worker, install the `@upstash/qstash`, a HTTP client to connect to your database QStash endpoint:
>     ```sh
>     $ npm install @upstash/qstash
>     ```
> 4. Check out an [example](https://docs.upstash.com/qstash/quickstarts/cloudflare-workers#3-use-qstash-in-your-handler) published by Upstash on how to receive webhooks from QStash in your Cloudflare Worker. 

Reviewer: ```suggestion
3. In your Worker, install the `@upstash/qstash`, a HTTP client to connect to your database QStash endpoint:
    ```sh
    $ npm install @upstash/qstash
    ```
4. Refer to the [Upstash documentation on how to receive webhooks from QStash in your Cloudflare Worker](https://docs.upstash.com/qstash/quickstarts/cloudflare-workers#3-use-qstash-in-your-handler). 
```
